---
title: "Commands for taxonomic analysis (16S rRNA gene and ITS2) of Pachnoda butana gut in two different suppliers"
output: html_notebook
---

```{r}
rm (list = ls())
```

# 16S

## Bacteria

```{r}
library(phyloseq)
physeqObjest <- import_biom("../qiime2/table-with-taxonomy-json2.biom",
                            refseqArgs=NULL, parseFunction = parse_taxonomy_default, 
                            parallel=FALSE)
```

```{r}
# METADATA
mapfile = read.csv("./metadata.csv", header = TRUE, sep=",")
samples = mapfile[,1]
rownames(mapfile) = samples
mapfile = mapfile[,2:length(colnames(mapfile))]

colnames(otu_table(physeqObjest))
rownames(mapfile)

ord = match(colnames(otu_table(physeqObjest)), rownames(mapfile), )

mapfile = mapfile[ord,]
sampledata = sample_data(mapfile)
physeq1 = merge_phyloseq(physeqObjest, sampledata)
sample_sums(physeq1)
```
```{r}
physeq1 = subset_taxa(physeq1, (Rank1 !="d__Eukaryota") | is.na(Rank1))
physeq1 = subset_taxa(physeq1, (Rank2 !="p__Vertebrata") | is.na(Rank2))
physeq1 = subset_taxa(physeq1, (Rank5 !="f__Mitochondria") | is.na(Rank5))
physeq1 = subset_taxa(physeq1, (Rank4 !="o__Chloroplast") | is.na(Rank4))
physeq1 = subset_taxa(physeq1, (Rank1 =="d__Bacteria"))
```

### PCoA

#### ASV level
```{r}
physeq1_rel  = transform_sample_counts(physeq1, function(x) x / sum(x)*100 )
sample_sums(physeq1_rel)
```

```{r}
library(ggrepel)
library(ggplot2)
brayDist <- phyloseq::distance(physeq1_rel, method="bray")
iMDS  <- ordinate(physeq1_rel)
p = p + geom_point(aes(size=3, alpha = 0.6))
bacteria <- plot_ordination(physeq_R6_rel, iMDS, color ="Supplier_GutArea") + theme_light() 
bacteria = bacteria + geom_point(aes(size=3, alpha = 0.6)) + scale_color_manual(values = c('#001f30', '#1e5775', '#5ca0c4', '#93b5c7', '#3b1401','#8c370b', '#c28261', '#d6b3a1'))
```

### AMPVIS2 (heatmap)

```{r}
#source the phyloseq_to_ampvis2() function from the gist
devtools::source_gist("8d0ca4206a66be7ff6d76fc4ab8e66c6")
colnames(tax_table(physeq_R6_rel)) = c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")
ampvis2_obj <- phyloseq_to_ampvis2(physeq_R6_rel)

rownames(ampvis2_obj$metadata) = ampvis2_obj$metadata[,"SampleName"]
colnames(ampvis2_obj$abund) = ampvis2_obj$metadata[,"SampleName"]
```

```{r}
ampvis2_obj
p = amp_heatmap(
      data = ampvis2_obj,
      facet_by = "Supplier_GutArea",
      group_by="SampleName",
      
      normalise = FALSE,
      tax_show = 20,
      tax_aggregate = "Genus",
      tax_add = "Family",
      plot_values_size = 3.4,
      min_abundance = 0.01,
      color_vector = c("royalblue3",
                   "whitesmoke",
                   "lightcoral"),
      round = 2
    )
```


## Archaea

```{r}
library(phyloseq)
physeqObjest <- import_biom("../qiime2/table-with-taxonomy-json2.biom",
                            refseqArgs=NULL, parseFunction = parse_taxonomy_default, 
                            parallel=FALSE)
```

```{r}
# METADATA
mapfile = read.csv("./metadata.csv", header = TRUE, sep=",")
samples = mapfile[,1]
rownames(mapfile) = samples
mapfile = mapfile[,2:length(colnames(mapfile))]

colnames(otu_table(physeqObjest))
rownames(mapfile)

ord = match(colnames(otu_table(physeqObjest)), rownames(mapfile), )

mapfile = mapfile[ord,]
sampledata = sample_data(mapfile)
physeq1 = merge_phyloseq(physeqObjest, sampledata)
sample_sums(physeq1)
```
```{r}
physeq1 = subset_taxa(physeq1, (Rank1 !="d__Eukaryota") | is.na(Rank1))
physeq1 = subset_taxa(physeq1, (Rank2 !="p__Vertebrata") | is.na(Rank2))
physeq1 = subset_taxa(physeq1, (Rank5 !="f__Mitochondria") | is.na(Rank5))
physeq1 = subset_taxa(physeq1, (Rank4 !="o__Chloroplast") | is.na(Rank4))
physeq1 = subset_taxa(physeq1, (Rank1 =="d__Archaea"))
```

### PCoA

#### ASV level
```{r}
physeq1_rel  = transform_sample_counts(physeq1, function(x) x / sum(x)*100 )
sample_sums(physeq1_rel)
```

```{r}
library(ggrepel)
library(ggplot2)
brayDist <- phyloseq::distance(physeq1_rel, method="bray")
iMDS  <- ordinate(physeq1_rel)
p = p + geom_point(aes(size=3, alpha = 0.6))
Archaea <- plot_ordination(physeq_R6_rel, iMDS, color ="Supplier_GutArea") + theme_light() 
Archaea = Archaea + geom_point(aes(size=3, alpha = 0.6)) + scale_color_manual(values = c('#001f30', '#1e5775', '#5ca0c4', '#93b5c7', '#3b1401','#8c370b', '#c28261', '#d6b3a1'))
```

### AMPVIS2 (heatmap)

```{r}
#source the phyloseq_to_ampvis2() function from the gist
devtools::source_gist("8d0ca4206a66be7ff6d76fc4ab8e66c6")
colnames(tax_table(physeq_R6_rel)) = c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")
ampvis2_obj <- phyloseq_to_ampvis2(physeq_R6_rel)

rownames(ampvis2_obj$metadata) = ampvis2_obj$metadata[,"SampleName"]
colnames(ampvis2_obj$abund) = ampvis2_obj$metadata[,"SampleName"]
```

```{r}
ampvis2_obj
p = amp_heatmap(
      data = ampvis2_obj,
      facet_by = "Supplier_GutArea",
      group_by="SampleName",
      
      normalise = FALSE,
      tax_show = 20,
      tax_aggregate = "Genus",
      tax_add = "Family",
      plot_values_size = 3.4,
      min_abundance = 0.01,
      color_vector = c("royalblue3",
                   "whitesmoke",
                   "lightcoral"),
      round = 2
    )
```

# ITS

## Fungi

```{r}
library(phyloseq)
physeqObjest <- import_biom("../qiime2/table-with-taxonomy-json2.biom",
                            refseqArgs=NULL, parseFunction = parse_taxonomy_default, 
                            parallel=FALSE)
```

```{r}
# METADATA
mapfile = read.csv("./metadata.csv", header = TRUE, sep=",")
samples = mapfile[,1]
rownames(mapfile) = samples
mapfile = mapfile[,2:length(colnames(mapfile))]

colnames(otu_table(physeqObjest))
rownames(mapfile)

ord = match(colnames(otu_table(physeqObjest)), rownames(mapfile), )

mapfile = mapfile[ord,]
sampledata = sample_data(mapfile)
physeq1 = merge_phyloseq(physeqObjest, sampledata)
sample_sums(physeq1)
```
```{r}
physeq1 = subset_taxa(physeq1, (Rank1 !="d__Eukaryota") | is.na(Rank1))
physeq1 = subset_taxa(physeq1, (Rank2 !="p__Vertebrata") | is.na(Rank2))
physeq1 = subset_taxa(physeq1, (Rank5 !="f__Mitochondria") | is.na(Rank5))
physeq1 = subset_taxa(physeq1, (Rank4 !="o__Chloroplast") | is.na(Rank4))
physeq1 = subset_taxa(physeq1, (Rank1 =="d__Fungi"))
```

### PCoA

#### ASV level
```{r}
physeq1_rel  = transform_sample_counts(physeq1, function(x) x / sum(x)*100 )
sample_sums(physeq1_rel)
```

```{r}
library(ggrepel)
library(ggplot2)
brayDist <- phyloseq::distance(physeq1_rel, method="bray")
iMDS  <- ordinate(physeq1_rel)
p = p + geom_point(aes(size=3, alpha = 0.6))
Fungi <- plot_ordination(physeq_R6_rel, iMDS, color ="Supplier_GutArea") + theme_light() 
Fungi = Fungi + geom_point(aes(size=3, alpha = 0.6)) + scale_color_manual(values = c('#001f30', '#1e5775', '#5ca0c4', '#93b5c7', '#3b1401','#8c370b', '#c28261', '#d6b3a1'))
```

### AMPVIS2 (heatmap)

```{r}
#source the phyloseq_to_ampvis2() function from the gist
devtools::source_gist("8d0ca4206a66be7ff6d76fc4ab8e66c6")
colnames(tax_table(physeq_R6_rel)) = c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")
ampvis2_obj <- phyloseq_to_ampvis2(physeq_R6_rel)

rownames(ampvis2_obj$metadata) = ampvis2_obj$metadata[,"SampleName"]
colnames(ampvis2_obj$abund) = ampvis2_obj$metadata[,"SampleName"]
```

```{r}
ampvis2_obj
p = amp_heatmap(
      data = ampvis2_obj,
      facet_by = "Supplier_GutArea",
      group_by="SampleName",
      
      normalise = FALSE,
      tax_show = 20,
      tax_aggregate = "Genus",
      tax_add = "Family",
      plot_values_size = 3.4,
      min_abundance = 0.01,
      color_vector = c("royalblue3",
                   "whitesmoke",
                   "lightcoral"),
      round = 2
    )

